### PR TITLE
Feature/external config merged with new api

### PR DESCRIPTION
**Pull request by zerginator**
Suggestion for external configurations in YAML ( use of ObjectMapper from Jackson)
further more introducing maven wrapper that every team member uses the same maven version

To use maven wrapper :
1. got to root directory of you project 
2. type " ./mvnw clean install"
3. Maven will be downloaded and runs current version 3.6.1

PS: i delete some duplicate dependencies as well
PPS: merged new-spotify-api

### CODE REVIEW

**Comment by zerginator** [comment]
Please have a special eye on Session.class , i refactored the inner class and outsourced the Blob encryption to another the common module

File: core/src/main/java/xyz/gianlu/librespot/cache/CacheManager.java [hunk 1]
```diff
@@ -29,21 +30,19 @@
     private static final long CLEAN_UP_THRESHOLD = TimeUnit.DAYS.toMillis(7);
     private static final Logger LOGGER = Logger.getLogger(CacheManager.class);
     private static final byte HEADER_TIMESTAMP = (byte) 0b11111111;
-    private final boolean enabled;
```
Comment (devgianlu) [inline core/src/main/java/xyz/gianlu/librespot/cache/CacheManager.java]: This breaks the entire cache system if it's disabled.

Comment (devgianlu) [reply core/src/main/java/xyz/gianlu/librespot/cache/CacheManager.java]: This will anyway cause an exception in `Session#cache()`.

File: common/pom.xml [hunk 2]
```diff
@@ -25,5 +29,24 @@
             <artifactId>gson</artifactId>
             <version>${gson.version}</version>
         </dependency>
+
+        <!-- Jackson dependency for parsing yaml files-->
```
Comment (devgianlu) [inline common/pom.xml]: Jackson is a bit on the heavy side for what it's being used, is it possible to switch to SnakeYAML which is much lighter?

File: common/pom.xml [hunk 3]
```diff
@@ -25,5 +29,24 @@
...
+
+        <dependency>
+            <groupId>com.fasterxml.jackson.core</groupId>
+            <artifactId>jackson-databind</artifactId>
+            <version>${jackson.version}</version>
+        </dependency>
+
+        <dependency>
+            <groupId>org.apache.commons</groupId>
+            <artifactId>commons-lang3</artifactId>
```
Comment (devgianlu) [inline common/pom.xml]: What's this for?

File: core/src/main/java/xyz/gianlu/librespot/cache/CacheManager.java (hunk 1 above)
Comment (zerginator) [reply core/src/main/java/xyz/gianlu/librespot/cache/CacheManager.java]: yes you are right , didn't see that connection there but its only used in 2 times in CDNManager and 1 time in AudioFileStreming and you are already handling null pointers at the most places. if i see correctly
image

Comment (zerginator) [reply core/src/main/java/xyz/gianlu/librespot/cache/CacheManager.java]: we could make CacheManager Optional and than handle it in the 2 Files ?

File: common/pom.xml (hunk 2 above)
Comment (zerginator) [reply common/pom.xml]: sure, ill try that

File: common/pom.xml (hunk 3 above)
Comment (zerginator) [reply common/pom.xml]: it was recommended for using Objectmapper

**Comment by crsmoro** [comment]
let me throw my 2 cents here..
i dont think that yaml is right answer for such a simple project..
librespot doesnt have that much parameters for a justification of "easy to read", it just going to let this bigger for (imo) no real improvement..
without the fatct that yaml is a bitc* with their space and tab things..

if you really wanna change to yaml i suggest to put this as an extra, and let the properties working too..

**Comment by devgianlu** [comment]
@crsmoro I think proprieties aren't very good either, what options are out here?
- JSON
- TOML
- YAML
- XML (overkill)
- Properties (looks bad with a lot of keys)

**Comment by zerginator** [comment]
I don't mind to use properties but since i havent seen them since years i thought it would be nice to have yml but json is also fine. Dont know about TOML. So up to you guys :-)

**Comment by crsmoro** [comment]
i know it's hard with the potions out there, but what we have today with properties is good enough, but that's me...
json like gianlu said, it's not designed for this..

**Comment by zerginator** [comment]
in second thought. yaml is quite good in mapping nested objects. like "player" object. when we stay with the properties we problaby have to read them in one by one and also have test all of them if they are empty and set default values. Would be a bit more boiler plate code. Or was there another way? @devgianlu

**Comment by devgianlu** [comment]
That's how it was before, default values were hardcoded inside the code and if the configuration file was missing something it would have been picked from `DefaultConfiguration`.

**Comment by zerginator** [comment]
yes, exactly my point , i could remove 3 files and have everything at one place

**Comment by devgianlu** [comment]
How? This PR adds 300 lines of code as of now.

**Comment by zerginator** [comment]
most of it is getters and setters...

**Comment by devgianlu** [comment]
Are the setters being used by `ObjectMapper`?

**Comment by zerginator** [comment]
yes they are used by ObjectMapper , but im not sure if snakeyml needs it ... but i guess thats the same

**Comment by devgianlu** [comment]
Anyway, I think I'll look into the TOML way. Or maybe we can support different configuration files. 

My thoughts also went over the ability to override any parameter from the command line which is not possible with this PR.

**Comment by zerginator** [comment]
ok , than we go with toml, not much difference as i see it : 
https://gist.github.com/oconnor663/9aeb4ed56394cb013a20
How do you want to use the command line? instant change when the application runs or just as an argument? In any case its possible to overwrite the configurations with the command line. just read the command line values in after you have read in the TOML configuration

**Comment by zerginator** [comment]
shall i look into TOML than ?

**Comment by devgianlu** [comment]
> How do you want to use the command line?

The same way it worked before, it would override the configuration file (if present). Like this:

    command line > conf file > default


> shall i look into TOML than ?

If I am going to change configuration file format, I'd switch to TOML only.

**Comment by devgianlu** [comment]
See #110
